### PR TITLE
Extend FESystem for more than a unique face

### PR DESCRIPTION
**Pull request by peterrum**
...references #12646

### CODE REVIEW

File: source/fe/fe_system.cc [hunk 1]
```diff
@@ -1409,7 +1409,9 @@ FESystem<dim, spacedim>::build_interface_constraints()
 {
   // TODO: the implementation makes the assumption that all faces have the
   // same number of dofs
-  AssertDimension(this->n_unique_faces(), 1);
+  if (this->n_unique_faces() > 1)
+    return;
```
Comment (peterrum) [inline source/fe/fe_system.cc:1413]: Not ideal but we currently don't support hanging-nodes for meshes with wedges/pyramids anyway.

File: include/deal.II/base/function.templates.h [hunk 2]
```diff
@@ -106,10 +106,9 @@ Function<dim, RangeNumberType>::vector_value_list(
 {
   // check whether component is in the valid range is up to the derived
   // class
-  Assert(values.size() == points.size(),
-         ExcDimensionMismatch(values.size(), points.size()));
+  // AssertDimension(points.size(), values.size());
 
-  for (unsigned int i = 0; i < points.size(); ++i)
+  for (unsigned int i = 0; i < values.size(); ++i)
```
Comment (bangerth) [inline include/deal.II/base/function.templates.h:111]: No, this isn't right. If the function is called with a different number of input and output arguments, then the place where the function is called needs to be fixed.

File: include/deal.II/fe/fe_base.h [hunk 3]
```diff
@@ -291,6 +291,13 @@ class FiniteElementData
    */
   const ReferenceCell reference_cell_kind;
 
+public:
+  /**
+   * GenericDoFsPerObject object this class was initialized with.
+   */
+  const internal::GenericDoFsPerObject data;
```
Comment (bangerth) [inline include/deal.II/fe/fe_base.h:298]: First, can we give this variable a better name? Second, why is this necessary?

Comment (bangerth) [reply include/deal.II/fe/fe_base.h:298]: Also, why is it `public`?

File: include/deal.II/fe/fe_tools.templates.h [hunk 4]
```diff
@@ -103,22 +100,78 @@ namespace FETools
...
-
-            multiplied_dofs_per_vertex +=
-              fes[i]->n_dofs_per_vertex() * multiplicities[i];
-            multiplied_dofs_per_line +=
-              fes[i]->n_dofs_per_line() * multiplicities[i];
-            multiplied_dofs_per_quad +=
-              fes[i]->n_dofs_per_quad(0) * multiplicities[i];
-            multiplied_dofs_per_hex +=
-              fes[i]->n_dofs_per_hex() * multiplicities[i];
+            if (initialized == false)
```
Comment (bangerth) [inline include/deal.II/fe/fe_tools.templates.h:108]: If I read this right, what you are doing here is initialize the dpo object the first time you find an element with `multiplicity>0`. I would much rather you do this before the current `for` loop, as in
```
  for (i=...)
    if (multiplicity>0)
      {
         ...do it here...
         break;
      }
```
I think that would be easier to read, and avoid the `initialized` variable altogether.

File: include/deal.II/fe/fe_tools.templates.h [hunk 5]
```diff
@@ -103,22 +100,78 @@ namespace FETools
...
+
+                initialized = true;
+              }
+
+            for (unsigned int j = 0;
+                 j < fes[i]->data.dofs_per_object_exclusive.size();
+                 ++j)
+              for (unsigned int k = 0;
+                   k < fes[i]->data.dofs_per_object_exclusive[j].size();
+                   ++k)
```
Comment (bangerth) [inline include/deal.II/fe/fe_tools.templates.h:148]: It seems to me like you're repeating many of these loop statements. Could their bodies be merged?

File: include/deal.II/grid/reference_cell.h [hunk 6]
```diff
@@ -1140,6 +1140,8 @@ ReferenceCell::face_to_cell_lines(const unsigned int  face,
   AssertIndexRange(face, n_faces());
   AssertIndexRange(line, face_reference_cell(face).n_lines());
 
+  static unsigned int X = numbers::invalid_unsigned_int;
```
Comment (bangerth) [inline include/deal.II/grid/reference_cell.h:1143]:
```suggestion
  static constexpr unsigned int X = numbers::invalid_unsigned_int;
```

File: include/deal.II/grid/reference_cell.h [hunk 7]
```diff
@@ -1176,11 +1178,25 @@ ReferenceCell::face_to_cell_lines(const unsigned int  face,
     }
   else if (*this == ReferenceCells::Pyramid)
     {
-      Assert(false, ExcNotImplemented());
+      const static ndarray<unsigned int, 5, 4> table = {{{{0, 1, 2, 3}},
```
Comment (bangerth) [inline include/deal.II/grid/reference_cell.h:1181]:
```suggestion
      static const ndarray<unsigned int, 5, 4> table = {{{{0, 1, 2, 3}},
```

File: include/deal.II/grid/reference_cell.h [hunk 8]
```diff
@@ -1140,6 +1140,8 @@ ReferenceCell::face_to_cell_lines(const unsigned int  face,
   AssertIndexRange(face, n_faces());
   AssertIndexRange(line, face_reference_cell(face).n_lines());
```
Comment (bangerth) [inline include/deal.II/grid/reference_cell.h:1141]: If you want, move the changes to this file into a separate PR. They're logically independent.

File: source/fe/fe_data.cc [hunk 9]
```diff
@@ -142,6 +142,7 @@ FiniteElementData<dim>::FiniteElementData(
   const Conformity                      conformity,
   const BlockIndices &                  block_indices)
   : reference_cell_kind(reference_cell)
+  , data(data)
```
Comment (bangerth) [inline source/fe/fe_data.cc:145]: From reading how the argument `data` is used below, the `data` member variable is duplicative of information we already store. I'm not in favor of that.

File: source/fe/fe_pyramid_p.cc [hunk 10]
```diff
@@ -103,6 +103,47 @@ FE_PyramidPoly<dim, spacedim>::FE_PyramidPoly(
       this->unit_support_points.emplace_back(-1.0, +1.0, 0.0);
       this->unit_support_points.emplace_back(+1.0, +1.0, 0.0);
       this->unit_support_points.emplace_back(+0.0, +0.0, 1.0);
+
+      this->unit_face_support_points.resize(5);
+
+      for (unsigned int f = 0; f < 5; ++f)
```
Comment (bangerth) [inline source/fe/fe_pyramid_p.cc:109]: Can we replace the hard-coded 5 in these two places by `ReferenceCells::Pyramid::n_faces()`? I think that would make the intent clearer.

File: source/fe/fe_pyramid_p.cc [hunk 11]
```diff
@@ -103,6 +103,47 @@ FE_PyramidPoly<dim, spacedim>::FE_PyramidPoly(
...
+}
+
+
+
+template <int dim, int spacedim>
+void
+FE_PyramidPoly<dim, spacedim>::
+  convert_generalized_support_point_values_to_dof_values(
+    const std::vector<Vector<double>> &support_point_values,
+    std::vector<double> &              nodal_values) const
```
Comment (bangerth) [inline source/fe/fe_pyramid_p.cc:135]: This function is also independently correct and we can merge that in a separate PR if you'd like.

File: source/fe/fe_system.cc (hunk 1 above)
Comment (bangerth) [reply source/fe/fe_system.cc:1413]: Hm, yes, not ideal. How can we catch this error in a better way? Is there a better place where this assertion could be put?

File: source/fe/fe_system.cc [hunk 12]
```diff
@@ -1682,7 +1684,7 @@ FESystem<dim, spacedim>::initialize(
 
     for (unsigned int face_no = 0; face_no < this->n_unique_faces(); ++face_no)
       {
-        this->face_system_to_component_table[0].resize(
+        this->face_system_to_component_table[face_no].resize(
```
Comment (bangerth) [inline source/fe/fe_system.cc:1687]: That was just a bug before, right?

File: source/fe/fe_wedge_p.cc [hunk 13]
```diff
@@ -103,14 +103,75 @@ FE_WedgePoly<dim, spacedim>::FE_WedgePoly(
 {
   AssertDimension(dim, 3);
 
-  if (degree == 1)
+  Assert(1 <= degree && degree <= 2, ExcNotImplemented())
+
+    for (double z = 0.0; z <= 1.0; z += (degree == 1 ? 1.0 : 0.5))
+  {
```
Comment (bangerth) [inline source/fe/fe_wedge_p.cc:109]: indentation looks funny here

File: source/fe/fe_wedge_p.cc [hunk 14]
```diff
@@ -103,14 +103,75 @@ FE_WedgePoly<dim, spacedim>::FE_WedgePoly(
 {
   AssertDimension(dim, 3);
 
-  if (degree == 1)
+  Assert(1 <= degree && degree <= 2, ExcNotImplemented())
+
+    for (double z = 0.0; z <= 1.0; z += (degree == 1 ? 1.0 : 0.5))
```
Comment (bangerth) [inline source/fe/fe_wedge_p.cc:108]:
```suggestion
    for (double z = 0.0; z <= 1.0; z += 1./degree)
```
This is going to run into trouble for values of `degree` for which `1./degree` is not exactly representable in floating point arithmetic. The test `z<=1` will then either work or not work, by accident. It's better to write this as
```
  for (unsigned int layer=0; layer<=degree; ++layer)
  {
     const double delta_z = 1./degree;
     const double z = layer * delta_z;
     ...
```

File: source/fe/fe_wedge_p.cc [hunk 15]
```diff
@@ -103,14 +103,75 @@ FE_WedgePoly<dim, spacedim>::FE_WedgePoly(
...
+  {
+    this->unit_support_points.emplace_back(0.0, 0.0, z);
+    this->unit_support_points.emplace_back(1.0, 0.0, z);
+    this->unit_support_points.emplace_back(0.0, 1.0, z);
+
+    if (degree == 2)
```
Comment (bangerth) [inline source/fe/fe_wedge_p.cc:114]: Maybe not for now, but if you wanted to leave a comment: These points are just the tensor product of the triangle points with `z`. We could simply create a `FE_SimplexP<2>(degree)` and ask for its support points; then you don't have to repeat the code for `degree>2` and it has to be implemented in only one place.

File: source/fe/fe_wedge_p.cc [hunk 16]
```diff
@@ -103,14 +103,75 @@ FE_WedgePoly<dim, spacedim>::FE_WedgePoly(
...
+      {
+        this->unit_support_points.emplace_back(0.5, 0.0, z);
+        this->unit_support_points.emplace_back(0.5, 0.5, z);
+        this->unit_support_points.emplace_back(0.0, 0.5, z);
+      }
+  }
+
+  this->unit_face_support_points.resize(5);
+
+  for (unsigned int f = 0; f < 5; ++f)
```
Comment (bangerth) [inline source/fe/fe_wedge_p.cc:124]: same thing here with the bare `5`

File: source/fe/fe_wedge_p.cc [hunk 17]
```diff
@@ -103,14 +103,75 @@ FE_WedgePoly<dim, spacedim>::FE_WedgePoly(
...
+        }
+    }
+}
+
+
+
+template <int dim, int spacedim>
+void
+FE_WedgePoly<dim, spacedim>::
+  convert_generalized_support_point_values_to_dof_values(
```
Comment (bangerth) [inline source/fe/fe_wedge_p.cc:161]: same here -- put it into a separate PR to make this one simpler.

**Review by bangerth — CHANGES_REQUESTED** [review]
